service/matching/matching_engine.go

```diff
@@ -2928,3 +2936,112 @@ func largerBacklogAge(rootBacklogAge *durationpb.Duration, currentPartitionAge *
+
+func prepareTaskQueueUserData(
+	tqud *persistencespb.TaskQueueUserData,
+	taskQueueType enumspb.TaskQueueType,
+) *persistencespb.TaskQueueUserData {
+	data := common.CloneProto(tqud)
+	if data == nil {
+		data = &persistencespb.TaskQueueUserData{}
+	}
+	if data.PerType == nil {
+		data.PerType = make(map[int32]*persistencespb.TaskQueueTypeUserData)
+	}
+	tqType := int32(taskQueueType)
+	if data.PerType[tqType] == nil {
+		data.PerType[tqType] = &persistencespb.TaskQueueTypeUserData{}
+	}
+	if data.PerType[tqType].Config == nil {
+		data.PerType[tqType].Config = &taskqueuepb.TaskQueueConfig{}
+	}
+	return data
+}
+
+func getTaskQueueConfig(
+	taskQueueUserData *persistencespb.TaskQueueUserData,
+	taskQueueType enumspb.TaskQueueType,
+) *taskqueuepb.TaskQueueConfig {
+	return taskQueueUserData.GetPerType()[int32(taskQueueType)].GetConfig()
+}
+
+func (e *matchingEngineImpl) UpdateTaskqueueConfig(ctx context.Context, request *matchingservice.UpdateTaskQueueConfigRequest) (*matchingservice.UpdateTaskQueueConfigResponse, error) {
+	taskQueueFamily, err := tqid.NewTaskQueueFamily(request.NamespaceId, request.UpdateTaskqueueConfig.GetTaskQueue())
+	if err != nil {
+		return nil, err
+	}
+	taskQueueType := request.UpdateTaskqueueConfig.GetTaskQueueType()
+	// Get the partition manager for the root workflow partition of the task queue family.
+	// Configuration updates are applied here and eventually propagate,
+	// to all partitions and associated activity task queues of the same task queue family.
+	tqm, _, err := e.getTaskQueuePartitionManager(ctx,
+		taskQueueFamily.TaskQueue(enumspb.TASK_QUEUE_TYPE_WORKFLOW).RootPartition(),
+		true, loadCauseOtherWrite)
+	if err != nil {
+		return nil, err
+	}
+	if request.GetUpdateTaskqueueConfig() == nil {
+		tqud, _, err := tqm.GetUserDataManager().GetUserData()
+		if err != nil {
+			return nil, err
+		}
+		// If no update is requested, return the current config.
+		return &matchingservice.UpdateTaskQueueConfigResponse{
+			UpdatedTaskqueueConfig: getTaskQueueConfig(tqud.GetData(), taskQueueType),
+		}, nil
+	}
+	updateOptions := UserDataUpdateOptions{Source: "UpdateTaskQueueConfig"}
+	_, err = tqm.GetUserDataManager().UpdateUserData(ctx, updateOptions,
+		func(tqud *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, bool, error) {
+			data := prepareTaskQueueUserData(tqud, taskQueueType)
+			// Update timestamp from hlc clock
+			existingClock := data.Clock
+			if existingClock == nil {
+				existingClock = hlc.Zero(e.clusterMeta.GetClusterID())
+			}
+			now := hlc.Next(existingClock, e.timeSource)
+			protoTs := hlc.ProtoTimestamp(now)
+			// Update relevant config fields
+			cfg := data.PerType[int32(taskQueueType)].Config
+			updateTaskQueueConfig := request.GetUpdateTaskqueueConfig()
+			// Queue Rate Limit
+			queueRateLimit := updateTaskQueueConfig.GetUpdateQueueRateLimit()
+			if queueRateLimit != nil {
+				cfg.QueueRateLimit = buildRateLimitConfig(queueRateLimit, protoTs)
+			}
```

Comment:
small suggestion: for values used if not-nil, declare and check in the same statement:
```suggestion
			if rl := updateTaskQueueConfig.GetUpdateQueueRateLimit(); rl != nil {
				cfg.QueueRateLimit = buildRateLimitConfig(rl, protoTs)
			}
```
and next one too

Reply:
Implemented for both QueueRateLimit and FairnessKeyRateLimit.